File: srclab-common/src/main/java/xyz/srclab/common/cache/CacheExpiry.java
package xyz.srclab.common.cache;

import xyz.srclab.annotation.Nullable;
import xyz.srclab.common.design.builder.ProductCachingBuilder;

import java.time.Duration;
import java.util.Objects;

/**
 * @author sunqian
 */
public interface CacheExpiry {

    static Builder newBuilder() {
        return new Builder();
    }

    @Nullable
    Duration expiryAfterCreate();

    @Nullable
    Duration expiryAfterRead();

    @Nullable
    Duration expiryAfterUpdate();

    final class Builder extends ProductCachingBuilder<CacheExpiry> {

        private @Nullable Duration expiryAfterCreate;
        private @Nullable Duration expiryAfterRead;
        private @Nullable Duration expiryAfterUpdate;

        public Builder expiryAfterCreate(Duration expiryAfterCreate) {
            this.expiryAfterCreate = expiryAfterCreate;
            this.updateState();
            return this;
        }

        public Builder expiryAfterRead(Duration expiryAfterRead) {
            this.expiryAfterRead = expiryAfterRead;
            this.updateState();
            return this;
        }

        public Builder expiryAfterUpdate(Duration expiryAfterUpdate) {
            this.expiryAfterUpdate = expiryAfterUpdate;
            this.updateState();
            return this;
        }

        @Override
        protected CacheExpiry buildNew() {
            return new CacheExpiryImpl(this);
        }

        private static final class CacheExpiryImpl implements CacheExpiry {

            private final @Nullable Duration expiryAfterCreate;
            private final @Nullable Duration expiryAfterRead;
            private final @Nullable Duration expiryAfterUpdate;

            private CacheExpiryImpl(Builder builder) {
                this.expiryAfterCreate = builder.expiryAfterCreate;
                this.expiryAfterRead = builder.expiryAfterRead;
                this.expiryAfterUpdate = builder.expiryAfterUpdate;
            }

            @Nullable
            @Override
            public Duration expiryAfterCreate() {
                return expiryAfterCreate;
            }

            @Nullable
            @Override
            public Duration expiryAfterRead() {
                return expiryAfterRead;
            }

            @Nullable
            @Override
            public Duration expiryAfterUpdate() {
                return expiryAfterUpdate;
            }

            @Override
            public boolean equals(Object object) {
                if (this == object) return true;
                if (object == null || getClass() != object.getClass()) return false;
                CacheExpiryImpl that = (CacheExpiryImpl) object;
                return Objects.equals(expiryAfterCreate, that.expiryAfterCreate) &&
                        Objects.equals(expiryAfterRead, that.expiryAfterRead) &&
                        Objects.equals(expiryAfterUpdate, that.expiryAfterUpdate);
            }

            @Override
            public int hashCode() {
                return Objects.hash(expiryAfterCreate, expiryAfterRead, expiryAfterUpdate);
            }

            @Override
            public String toString() {
                return "(" +
                        "expiryAfterCreate=" + expiryAfterCreate +
                        ", expiryAfterRead=" + expiryAfterRead +
                        ", expiryAfterUpdate=" + expiryAfterUpdate +
                        ')';
            }
        }
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/cache/CacheExpiry.kt
package xyz.srclab.common.cache

import xyz.srclab.common.base.CachingProductBuilder
import java.time.Duration

/**
 * @author sunqian
 */
interface CacheExpiry {

    fun expiryAfterCreate(): Duration
    fun expiryAfterRead(): Duration
    fun expiryAfterUpdate(): Duration

    class Builder : CachingProductBuilder<CacheExpiry>() {

        private var expiryAfterCreate: Duration = Duration.ZERO
        private var expiryAfterRead: Duration = Duration.ZERO
        private var expiryAfterUpdate: Duration = Duration.ZERO

        fun expiryAfterCreate(expiryAfterCreate: Duration): Builder {
            this.expiryAfterCreate = expiryAfterCreate
            this.commitChange()
            return this
        }

        fun expiryAfterRead(expiryAfterRead: Duration): Builder {
            this.expiryAfterRead = expiryAfterRead
            this.commitChange()
            return this
        }

        fun expiryAfterUpdate(expiryAfterUpdate: Duration): Builder {
            this.expiryAfterUpdate = expiryAfterUpdate
            this.commitChange()
            return this
        }

        override fun buildNew(): CacheExpiry {
            return CacheExpiryImpl(this)
        }

        private class CacheExpiryImpl(builder: Builder) : CacheExpiry {

            private val expiryAfterCreate: Duration
            private val expiryAfterRead: Duration
            private val expiryAfterUpdate: Duration

            override fun expiryAfterCreate(): Duration {
                return expiryAfterCreate
            }

            override fun expiryAfterRead(): Duration {
                return expiryAfterRead
            }

            override fun expiryAfterUpdate(): Duration {
                return expiryAfterUpdate
            }

            override fun equals(other: Any?): Boolean {
                if (this === other) return true
                if (javaClass != other?.javaClass) return false

                other as CacheExpiryImpl

                if (expiryAfterCreate != other.expiryAfterCreate) return false
                if (expiryAfterRead != other.expiryAfterRead) return false
                if (expiryAfterUpdate != other.expiryAfterUpdate) return false

                return true
            }

            override fun hashCode(): Int {
                var result = expiryAfterCreate.hashCode() ?: 0
                result = 31 * result + (expiryAfterRead.hashCode() ?: 0)
                result = 31 * result + (expiryAfterUpdate.hashCode() ?: 0)
                return result
            }

            init {
                expiryAfterCreate = builder.expiryAfterCreate
                expiryAfterRead = builder.expiryAfterRead
                expiryAfterUpdate = builder.expiryAfterUpdate
            }
        }
    }

    companion object {

        @JvmField
        val ZERO = newBuilder().build()

        @JvmStatic
        fun newBuilder(): Builder {
            return Builder()
        }
    }
}